### PR TITLE
Refactors all the things

### PR DESCRIPTION
**Pull request by t6d**
@christhomson, please review if you have a moment.

### CODE REVIEW

File: lib/smart_properties.rb [hunk 1]
```diff
@@ -31,13 +31,7 @@ module ClassMethods
     # @return [Hash<String, Property>] A map of property names to property instances.
     #
     def properties
-      @_smart_properties ||= begin
-        parent = if self != SmartProperties
```
Comment (christhomson) [inline lib/smart_properties.rb]: Is the `self != SmartProperties` condition not needed anymore?

**Comment by christhomson** [comment]
Just one question, otherwise :ship: :it:

**Comment by t6d** [comment]
@kevinhughes27, if you have a moment, feel free to take a look.

File: lib/smart_properties.rb (hunk 1 above)
Comment (t6d) [inline lib/smart_properties.rb]: It is not. The check is performed in `PropertyCollection.for` when the ancestor tree traversing is done.

File: lib/smart_properties/property.rb [hunk 2]
```diff
@@ -79,10 +90,27 @@ def define(klass)
 
       scope.send(:attr_reader, name)
       scope.send(:define_method, :"#{name}=") do |value|
-        instance_variable_set("@#{property.name}", property.prepare(value, self))
+        instance_variable_set("@#{property.name}", property.prepare(self, value))
```
Comment (benedikt) [inline lib/smart_properties/property.rb]: Shouldn't this use `set` instead? Or at least `instance_variable_name`?

File: lib/smart_properties.rb (hunk 1 above)
Comment (christhomson) [inline lib/smart_properties.rb]: :+1:

File: lib/smart_properties/property.rb (hunk 2 above)
Comment (t6d) [inline lib/smart_properties/property.rb]: Nice catch! :)

**Pull request merged**